Introduced option to make parser more lenient regarding unknown entry types

johangrufman opened the pull request:
I had some trouble parsing a SIE file with entries like `#MOMSKOD 2611 10`. When parsing such a file I get

    key not found: "momskod"
    sie (2.1.0) lib/sie/parser/tokenizer/token.rb:14:in `fetch'
    sie (2.1.0) lib/sie/parser/tokenizer/token.rb:14:in `entry_type'
    sie (2.1.0) lib/sie/parser/line_parser.rb:15:in `parse'
    sie (2.1.0) lib/sie/parser.rb:37:in `parse_line'
    sie (2.1.0) lib/sie/parser.rb:27:in `block in parse'
    sie (2.1.0) lib/sie/parser.rb:17:in `each_line'
    sie (2.1.0) lib/sie/parser.rb:17:in `parse'

This is understandable because as far as I can tell #MOMSKOD is not mentioned in the spec. The file I tried to parse is exported from iOrdning and I don't know if MOMSKOD is some custom extension or if it's from an older version of the spec. 

Anyway, I still need to be able to parse files like these so I tried to introduce an option to make the parser more lenient regarding unknown entry types. Given `lenient: true` the parser now handles unknown entry types without raising any error. Without the option it still raises an error. See the updated specs for details.

There might be things to discuss regarding the implementation but first of all I'd like to know if you'd consider such an option useful.

henrik commented:
Sounds like a good thing to have, but I'll leave this to @joakimk since he's done most of the work on this lib.

I wonder if something more explicit than "lenient" would be good. "raise_on_unknown_entries"? I assume you'll write it maybe once for an entire project, and read it a thousand times, so the clearer the better maybe, even if it's longer.

henrik commented:
Oh, and thanks so much for making a pull request! :tada:

johangrufman commented:
Well, thank *you* for open sourcing this gem in the first place. :-) 

The option name "lenient" was by inspired by similar options I've seen in other libraries. See for example [the DateFormat class in Java](http://docs.oracle.com/javase/6/docs/api/java/text/DateFormat.html#setLenient%28boolean%29). My first thought was that there might be other non standard things the parser might encounter in an exported SIE file and then the option "lenient" could then encompass them all (rather than having multiple raise_on_xyz options).

That said, I'm perfectly fine with a more explicit option name too. Just tell me what you prefer.

tskogberg commented:
What about `accept_non_standardised_entries`?

henrik commented:
@johangrufman Aha, that makes some sense.

I guess it should depend on why this option is needed at all.

In your case, you obviously don't want it because your source doesn't seem to stick to the core entry types. Would a whitelist make sense then, where you could specify that it should also accept "momskod" and "foobar" but nothing else? Is it valuable to have the lib catch unknown entry types at all, as opposed to just always being lenient for everyone?

If there *is* good reason to have an option, then the question is whether the same reason would apply to any future "leniency". Is it basically "we have standard files so some extra validations are nice" vs. "we have weird files so there can be all kinds of oddities"? If so, a single "lenient" option probably does make sense.

henrik commented:
@tskogberg Hm, yeah, something like that could also work. "accept_non_standard_entries" maybe. If that's the right terminology – I don't know the spec. Is "momskod" a "non-standard entry"?

Koronen commented:
I can confirm that `MOMSKOD` is not a recognized entry in the SIE format specification (4B edition).

johangrufman commented:
@henrik I guess that whether a whitelist makes sense depends a bit on the culture around these files. If it's common for vendors to introduce custom entries in the files I think it would be inconvenient to maintain a whitelist. I'd like my application to be able to continue parsing the files even if some vendor introduces another custom entry type in the future. 

Whether the parser should be lenient by default is also a good question. I think that if I'd developed this for my own purposes only, I would have skipped the option and made the parser lenient by default. Basically, the only time I'd like the parser to raise an error would be if the given file is not a SIE file at all (if the user by mistake uploads a word file for example). In case it really is a SIE file and I need to know if the file contains unknown entries I think I'd prefer to query the parse result afterwords. Something like:
```ruby
    parser = Sie::Parser.new
    sie_file = parser.parse(file)
    if sie_file.has_unknown_entries?
      # Do something
    end

henrik commented:
My takeaway from this discussion is that a single boolean option is a simple thing that would work for now – a binary "validate against spec, or not" option. I guess something more broad like "lenient" could be a good idea in case we have more things we'll need to loosen.

As long as it's documented, I could get behind any of the names suggested.

One more thing I'd suggest: make the error message helpfully suggest toggling this option.

joakimk commented on lib/sie/parser.rb:
```diff
@@ -33,8 +37,14 @@ def parse(io)
 
     private
 
+    attr_reader :options
```
We have https://github.com/barsoom/attr_extras as a dependency for this library so you can do "attr_private :options" up beside initialize.

johangrufman replied:
Just to make sure I understand before I update: You prefer `attr_private :options`next to initialize rather than `attr_reader :options` among the other private methods?

joakimk replied:
Yep https://github.com/barsoom/attr_extras#why-not-use-private-attr_reader-foo-instead-of-attr_private-foo

joakimk commented on lib/sie/parser/line_parser.rb:
```diff
@@ -5,13 +5,29 @@
   class Parser
     class LineParser
-      pattr_initialize :line
+      pattr_initialize :line, [:lenient]
```
Not terribly important, but we generally do `[ :lenient ]`. I see we haven't been consistent in this library though.

joakimk commented on lib/sie/parser/line_parser.rb:
```diff
@@ -5,13 +5,29 @@
         first_token = tokens.shift
 
         entry = Entry.new(first_token.label)
+        if first_token.known_entry_type?
```
Whitespace over this line would be nice :)

joakimk commented on spec/integration/parser_spec.rb:
```diff
@@ -18,7 +18,39 @@
+  context 'unknown entries' do
+    let(:file_with_unknown_entries) { "fixtures/sie_file_with_unknown_entries.se" }
+
+    context 'lenient parser' do
```
Try to keep a consistent quote style, we prefer `"` unless there is a special reason to use `'`.

joakimk replied:
This won't read very well as a spec name, it will be "unknown entries lenient parser handles unknown..". Perhaps "with unknown entries", "using a lenient parser".

joakimk commented on spec/integration/parser_spec.rb:
```diff
@@ -18,7 +18,39 @@
+    end
+
+    context 'rigorous parser' do
+      let(:parser) {Sie::Parser.new}
```
Inconsistent style, we prefer `{ Sie... }`. I see you have written in both ways in different files.

joakimk commented on spec/unit/parser/line_parser_spec.rb:
```diff
@@ -14,6 +14,26 @@
+        expect { parser.parse }.to raise_error
+      end
+    end
+  end
```
I wonder if we need an integration test if we cover the difference here?

johangrufman replied:
If it had been only up to me, I'd keep the integration test anyway. It still covers the communication between Parser and LineParser.

joakimk replied:
Oh, I see now. Sure, keep it.

joakimk commented:
I think an option to make the parser more lenient could be a good idea. I'd like the code to be a bit more consistent though and the README to mention the feature.

Also, as henrik said, it would be helpful if the error message includes a hint about the feature.

johangrufman commented:
I have now pushed the following updates:
* More consistent code style. Hopefully, I have now fixed the things mentioned in the comments. Please tell me if I misunderstood something. (I understand the value of consistency and I don't mind fixing if something could be improved.)
* The error message now mentions the lenient option
* The README now includes information about the lenient option

johangrufman commented:
@henrik I fixed the code styling of "lenient".

Yes, the nested conditional bothered me too. And for some irrational reason it bothers me more when I changed it from
```ruby
raise "Error" unless lenient
````
to
```ruby
unless lenient
  raise "Error"
end
```
I guess a more complete version of your suggestion would be something like
```ruby
    def parse
        tokens = tokenize(line)
        first_token = tokens.shift

        verify_entry_type(first_token)

        entry = Entry.new(first_token.label)

        if first_token.known_entry_type?
          build_entry(entry, first_token, tokens)
        end

        entry
    end

    private

    def verify_entry_type(token)
        unless token.known_entry_type? || lenient
          raise "Unknown entry type: #{token.label}. Pass 'lenient: true' to Parser.new to avoid exception."
        end
    end
```
The only thing that bothers me with this version is that the condition `token.known_entry_type?` appears twice. Maybe it's still better though. I guess an early exit from the method could be an alternative too but I don't quite see how that would improve the method.

johangrufman commented:
Any news on this? Is there anything else you want me to fix? If you prefer the version of LineParser in my last comment with less nested conditionals (but with `token.known_entry_type?` appearing twice) I can fix that too. Just tell me what you prefer.

henrik commented:
@joakimk Just to be explicit, I consider this to be "your" PR since it's in the sie repo.

joakimk commented:
I merged the branch and fixed a few things in https://github.com/barsoom/sie/commit/f5de2a28cdc46750d4801acd83f58f408356b373. One important thing to learn from that code is that we should probably never use "to raise_error" without arguments in tests.

Thanks for the contribution. I've released it as 2.2.0.
